lib/task/import/validate/csvParentValidator.class.php

```diff
@@ -0,0 +1,214 @@
+{
+    const TITLE = 'Parent check';
+    // Persist across multiple CSVs.
+    protected $legacyIdList = [];
+
+    // Reset after every CSV.
+    protected $orphanRowsFound = false;
+    protected $parentIdColumnPresent;
+    protected $qubitParentSlugColumnPresent;
+    protected $legacyIdColumnPresent;
+    protected $unmatchedCount = 0;
+    protected $rowsWithParentId = 0;
+    protected $rowsWithQubitParentSlug = 0;
+    protected $rowsWithParentIdQubitParentSlug = 0;
+    protected $rowsWithoutParentIdQubitParentSlug = 0;
+
+    public function __construct(array $options = null)
+    {
+        $this->setTitle(self::TITLE);
+
+        parent::__construct($options);
+    }
+
+    public function reset()
+    {
+        $this->orphanRowsFound = false;
+        $this->parentIdColumnPresent = null;
+        $this->qubitParentSlugColumnPresent = null;
+        $this->legacyIdColumnPresent = null;
+        $this->unmatchedCount = 0;
+        $this->rowsWithParentId = 0;
+        $this->rowsWithQubitParentSlug = 0;
+        $this->rowsWithParentIdQubitParentSlug = 0;
+        $this->rowsWithoutParentIdQubitParentSlug = 0;
+
+        parent::reset();
+    }
+
+    public function testRow(array $header, array $row)
+    {
+        parent::testRow($header, $row);
+        $row = $this->combineRow($header, $row);
+
+        // Is qubitParentSlug column present?
+        if (!isset($this->qubitParentSlugColumnPresent)) {
+            $this->qubitParentSlugColumnPresent = array_key_exists('qubitParentSlug', $row);
+        }
+
+        // Is parentId column present?
+        if (!isset($this->parentIdColumnPresent)) {
+            $this->parentIdColumnPresent = array_key_exists('parentId', $row);
+        }
+
+        // Is legacyId column present?
+        if (!isset($this->legacyIdColumnPresent)) {
+            $this->legacyIdColumnPresent = array_key_exists('legacyId', $row);
+        }
+
+        // When both are present, qubitParentSlug will override parentId.
+        if ($this->qubitParentSlugColumnPresent && !empty($row['qubitParentSlug'])) {
+            ++$this->rowsWithQubitParentSlug;
+
+            if (!$this->canFindBySlug($row['qubitParentSlug'], $this->options['className'])) {
+                $this->orphanRowsFound = true;
+                ++$this->unmatchedCount;
+
+                // Add row that triggered this to the output.
+                $this->testData->addDetail(implode(',', $row));
+            }
+        } elseif ($this->parentIdColumnPresent && !empty($row['parentId'])) {
+            ++$this->rowsWithParentId;
+
+            if (!$this->canFindByParentId($row['parentId'], $this->options['source'])) {
+                $this->orphanRowsFound = true;
+                ++$this->unmatchedCount;
+
+                // Add row that triggered this to the output.
+                $this->testData->addDetail(implode(',', $row));
+            }
+        }
+
+        if ($this->parentIdColumnPresent && empty($row['parentId'])
```

Comment:
I don't know what the "canonical" formatting is for multi-line conditionals like this, but my preference is:

```
if (
     $this->parentIdColumnPresent 
     && empty($row['parentId'])
     && $this->qubitParentSlugColumnPresent
     && empty($row['qubitParentSlug'])
) {
```
            ++$this->rowsWithoutParentIdQubitParentSlug;
        }

Reply:
Confirmed that my suggestion is canonical: https://wiki.accesstomemory.org/wiki/Development/Coding_standard#Multi-line_control_structures

Reply:
Nice - TIL...

Reply:
Completed in this file. Will look around for other instances.